passes/techmap/clock_gate.cc

```diff
@@ -0,0 +1,205 @@
+#include "kernel/yosys.h"
+#include "kernel/ff.h"
+#include <optional>
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct ClockGateCell {
+	std::string name;
+	std::string ce_pin;
+	std::string clk_in_pin;
+	std::string clk_out_pin;
+
+	ClockGateCell(std::string& name, std::string& str) : name(RTLIL::escape_id(name)) {
+		char delimiter = ':';
+		size_t pos1 = str.find(delimiter);
+		if (pos1 == std::string::npos)
+			log_assert(false && "Not enough ports in descriptor string");
+		size_t pos2 = str.find(delimiter, pos1 + 1);
+		if (pos2 == std::string::npos)
+			log_assert(false && "Not enough ports in descriptor string");
+		size_t pos3 = str.find(delimiter, pos2 + 1);
+		if (pos3 != std::string::npos)
+			log_assert(false && "Too many ports in descriptor string");
+
+		ce_pin = str.substr(0, pos1);
+		ce_pin = RTLIL::escape_id(ce_pin);
+
+		clk_in_pin = str.substr(pos1 + 1, pos2 - (pos1 + 1));
+		clk_in_pin = RTLIL::escape_id(clk_in_pin);
+
+		clk_out_pin = str.substr(pos2 + 1, str.size() - (pos2 + 1));
+		clk_out_pin = RTLIL::escape_id(clk_out_pin);
+	}
+};
+
+struct ClockgatePass : public Pass {
+	ClockgatePass() : Pass("clock_gate", "extract clock gating out of flip flops") { }
+	void help() override {
+		// TODO
+	}
+
+	SigMap sigmap;
+	FfInitVals initvals;
+
+	// One ICG will be generated per ClkNetInfo
+	// if the number of FFs associated with it is sufficent
+	struct ClkNetInfo {
+		// Original, ungated clock into enabled FF
+		Wire* clk_net;
+		// Original clock enable into enabled FF
+		Wire* ce_net;
+		bool pol_clk;
+		bool pol_ce;
+		unsigned int hash() const {
+			unsigned int h = mkhash_init;
+			h = mkhash(h, hash_ptr_ops::hash(clk_net));
+			h = mkhash(h, hash_ptr_ops::hash(ce_net));
+			h = mkhash(h, pol_clk);
+			h = mkhash(h, pol_ce);
+			return h;
+		}
+		bool operator==(const ClkNetInfo& other) const {
+			return (clk_net == other.clk_net) &&
+			       (ce_net == other.ce_net) &&
+			       (pol_clk == other.pol_clk) &&
+			       (pol_ce == other.pol_ce);
+		}
+	};
+
+	struct GClkNetInfo {
+		// How many CE FFs on this CLK net have we seen?
+		int net_size;
+		// After ICG generation, we have new gated CLK signals
+		Wire* new_net;
+	};
+
+	ClkNetInfo clk_info_from_ff(FfData& ff) {
+		Wire* clk = ff.sig_clk.as_wire();
+		Wire* ce = ff.sig_ce.as_wire();
```

Comment:
There's no guarantee these two are single-bit wires (when not the `as_wire()` conversion will crash). The solution is for the `ClkNetInfo` fields to be sigbits

Reply:
I considered this but forgot to put down a TODO huh